### PR TITLE
feature/83-authentication-system

### PR DESCRIPTION
**Pull request by Ahydul**
The login endpoint is /api/login. It doesn't appear visually because the login coudn't be implemented using the routers.

This PR introduces some changes to the settings of the proyect. Now the user used by the application is our custom User and not Django's. To create a superuser you need to provide an email and a password. If you are using a script to create the superuser take into account that it no longer wants an username. 

This is a valid script to makemigrations, migrate and createsuperuser:

python manage.py makemigrations nexong && 
python manage.py migrate && 
export DJANGO_SUPERUSER_PASSWORD='admin' &&
python manage.py createsuperuser --email admin@admin.com --noinput

### CODE REVIEW

**Review by FelixoGudiel — DISMISSED** [review]
Although the functionality is correct, I have found one problem and some doubts have arisen:
-As you have changed the user model, the populate does not work. Please fix it.
Now, about doubts:
-I believe that the work that you have done in hashing passwords (right now it only works in createsuperuser) should also be done for regular creation of user, but I dont know if it should be another task on its own.
-You have added two new attributes to user (is_admin and is_staff). Are these completely necessary? Our requirements do not need them, but if it is necessary for the auth, I would understand it.
If anyone involved in this issue could give me their take on these doubts I would thank it.

**Comment by Ahydul** [comment]
> right now it only works in createsuperuser) should also be done for regular creation of user, but I dont know if it should be another task on its own

I haven't done anything to the passwords. The application hashes them by default, the only change I've done is that now the default User used is our own. For passwords to be hashed it needs to be modified in the user serializer I believe and the model may need to change the password in some way. That's outside the scope o this issue though.

> You have added two new attributes to user (is_admin and is_staff). Are these completely necessary? Our requirements do not need them, but if it is necessary for the auth, I would understand it.

The application fails if you don't provide them and I don't know any way to circumvent this. If anyone has an idea let me know.

File: nexong/models.py [hunk 1]
```diff
@@ -194,7 +194,34 @@ class Educator(models.Model):
...
+        user = self.model(email=self.normalize_email(email))
+        user.set_password(password)
+        user.is_admin = False
+        user.is_staff = False
+        user.save(using=self._db)
+        return user
+
+    def create_superuser(self, email, password=None):
+        user = self.create_user(email, password=password)
+        user.is_admin = True
```
Comment (pabpercab1) [inline nexong/models.py:213]: At the moment, for me says:
`dajngo.db.utils.ProgrammingError: Column nexong_user.is_admin does not exist`
Probably the problem is related to the line highlighted.

**Review by pabpercab1 — CHANGES_REQUESTED** [review]
My suggestion is to try to implement an existing framework with the authorisation features. There exist comprehensive manuals and guides to achieve that. Check some of these:
- https://www.django-rest-framework.org/api-guide/authentication/#django-oauth-toolkit
- https://github.com/jazzband/django-oauth-toolkit
- https://jpadilla.github.io/django-rest-framework-oauth/
- https://auth0.com/docs/quickstart/backend/django/01-authorization

**Comment by Ahydul** [comment]
> My suggestion is to try to implement an existing framework with the authorisation features

Those are oauth authentication implementations and that's outside the scope of this issue. I believe they can be integrated in the future with my token authentication implementation

File: nexong/models.py (hunk 1 above)
Comment (Ahydul) [reply nexong/models.py:213]: That is strange, one of the latest changes broke the app. Thanks for pointing that out.

Comment (Ahydul) [reply nexong/models.py:213]: Ok nevermind, if you reset the database (deleting the migrations) it works as expected.

**Review by pabpercab1 — CHANGES_REQUESTED** [review]
Your functionality works as expected. However, I think in the same way as my colleague, the fields `is_admin` and `is_staff`, overwrites the current function of the role requested by the organisation. Try to implement it using another method that keeps the functionality of the role in the user model.

**Comment by Ahydul** [comment]
> Your functionality works as expected. However, I think in the same way as my colleague, the fields `is_admin` and `is_staff`, overwrites the current function of the role requested by the organisation. Try to implement it using another method that keeps the functionality of the role in the user model.

As I said previously, I don't know a way of doing that. Those two parameters are required by Django. The only reason I'm using them is because we are implementing the User model completely by using AbstractBaseUser instead of AbstractUser, which wuold implement them by default. I would recommend not using the roles either way.

**Review by pabpercab1 — COMMENTED** [review]
Maybe you can check this link to create a custom auth:

[Customizing auth](https://docs.djangoproject.com/en/5.0/topics/auth/customizing/)

**Pull request closed without merging**